### PR TITLE
Transformations: De-emphasize non-applicable transformations

### LINKED ISSUE #74199: Transformations: Make it harder to select unusable transformations from UI
Currently all transformations will be shown in the UI regardless of the underlying data. This can end up presenting confusing situations, especially if someone isn't aware of the underlying operation of transformations.

For instance, the _Merge_ transformation requires multiple data frames. However it can still be added even when there is only a single frame:

Screenshot 2023-08-31 at 10 13 41 AM

We should apply a method to hide or disable these in appropriate situations. In the basic case with a single data series _Merge_ can never be applied so it's a bit of a shame we present that as an option. Another good candidate would be _Format Time_ as it requires a time field to function.

### PR DESCRIPTION
**Pull request by codeincarnate**
**What is this feature?**

This PR adds support to transformations for an `applicator` function which can be used to determine if the transformation is applicable to the current data.

**Why do we need this feature?**

Many transformations have base conditions needed for them to work correctly. Currently however, we don't do any checks to see if these base conditions are met. This makes for a confusing experience, as it's currently only possible to tell through experimentation or from prior knowledge what given transformations will work on a given data set.

**Who is this feature for?**

Users of transformations

**Which issue(s) does this PR fix?**:

Fixes #74199

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by codeincarnate** [comment]
Oops, forgot to put this in draft 😢. Still needs a little to hammer out.

**Comment by joshhunt** [comment]
neat! 📸

**Comment by oscarkilhed** [comment]
Nice! Will review in the morning!

File: packages/grafana-data/src/types/transformations.ts [hunk 1]
```diff
@@ -28,6 +28,12 @@ export interface DataTransformerInfo<TOptions = any> extends RegistryItemWithOpt
    * @param options
    */
   operator: (options: TOptions, context: DataTransformContext) => MonoTypeOperatorFunction<DataFrame[]>;
+  /**
+   * Function that is present will indicate whether a transformation is applicable 
+   * given the current data.
+   * @param options
+   */
+  applicator?: (data: DataFrame[]) => boolean;
```
Comment (ryantxu) [inline packages/grafana-data/src/types/transformations.ts:36]:
```suggestion
  isApplicable?: (data: DataFrame[]) => boolean;
```

perhaps too complicated, but we could also consider something like panel suggestions.  where each transformation is passed the data and we get a score+possible config options from it 🤔 

for the minimum version, returning a low score is the same as returning false here

Comment (codeincarnate) [reply packages/grafana-data/src/types/transformations.ts:36]: I'm very much for this! I'll see what I can do for an MVP 😄

Comment (codeincarnate) [reply packages/grafana-data/src/types/transformations.ts]: Pushed updates which should resolve these

**Comment by codeincarnate** [comment]
Added a few things:
- Moved to applicability levels. Under the hood there are four right now, but we reduce that to two. (i.e. we ignore the impossible outcome and the highly applicable outcome)
- Adds HighlyApplicable, which would allow transformations to be highlighted in the future based on data (i.e. promote heatmap if heatmap data is found)
- Updates the display to show a grayscale icon if disabled, add disabled to the title, and to add an info icon which contains the applicability message to help the user understand the source of the error.

Currently although they're marked as disabled they can still be used as we don't wish to make these unusable for cases that we haven't taken into account.

**Comment by codeincarnate** [comment]
@ryantxu if you could take another look, I moved to the applicability levels per your suggestion and I think we can get quite a lot from this! Would be great to know your thoughts in regard to the implementation there.

**Review by oscarkilhed — COMMENTED** [review]
This is looking good!
There's a few other transformations that don't really work with all kinds of data. TimeSeriesToTable for instance, doesn't work with multiple fields in the same series. But adding and tuning these rules could be left for follow up PRs i guess.

**Comment by oscarkilhed** [comment]
Two thoughts.
Maybe it's a good idea to sort the disabled transformations to the bottom. I also wouldn't mind if they were a bit more de-emphasized, the gray-scale on the icons is nice, but it's not super obvious.

The transformations are still selectable. I'm a bit torn on this, I think right now they need to be enabled, because there is no way to insert a transformation between two transformations, a transformation could be applicable before one of the applied transformations, but not after it. But it's also a bit weird having a transformation say that it's disabled but you can still apply it.

**Comment by codeincarnate** [comment]
Thanks for the comments Oscar!

1. In terms of the conditions, I was hoping to get a decent base set. I think we can go pretty arbitrarily deep in this so didn't go crazy for this PR.
2. In terms of the sorting to the bottom, there was some concern that they would end up being somewhat hidden. Given that we are pushing to move to a drawer interface for new transformations though this may be ok. I'll circle back with @catherineymgui  about that.
3. I'll play with some ways to make things look less uhhh selectable 😆

**Comment by codeincarnate** [comment]
@mdvictor @oscarkilhed if you could take another look at the presentation it could be appreciated 🙏

**Review by oscarkilhed — APPROVED** [review]
Looks good! :tada:

**Pull request merged**